### PR TITLE
feat: Retry on error

### PR DESCRIPTION
**Pull request by alex-barstow**
## Description
This adds support for a `retryOnError` option that will have the player select a new source from the source list when the selected one fails prior to playback. If a source fails during playback, no retry will be attempted.

### CODE REVIEW

File: src/js/player.js [hunk 1]
```diff
@@ -3377,6 +3378,26 @@ class Player extends Component {
...
       middleware.setTech(mws, this.tech_);
     });
+
+    // Try another available source if this one fails before playback.
+    if (this.options_.retryOnError && sources.length > 1) {
+      const retry = () => {
+        // Remove the error modal
+        this.error(null);
+        this.retrying_ = true;
+        this.src(sources.slice(1));
```
[the remark this answers is not in the document]
Comment (gkatsev) [reply src/js/player.js:3388]: No, I think that piece only is actually badly named. The `err` there is actually whether any techs can play the currently chosen source. So, this still makes sense. I think the other piece would trigger if we were to add say an RTMP source. Though, that's actually a good thing to verify that the two pieces don't step on each other's toes.

**Review by gkatsev — COMMENTED** [review]
Not sure why tests are failing, I would ignore that for now.

File: src/js/player.js [hunk 2]
```diff
@@ -3377,6 +3378,26 @@ class Player extends Component {
 
       middleware.setTech(mws, this.tech_);
     });
+
+    // Try another available source if this one fails before playback.
+    if (this.options_.retryOnError && sources.length > 1) {
+      const retry = () => {
+        // Remove the error modal
+        this.error(null);
+        this.retrying_ = true;
```
[the remark this answers is not in the document]
Comment (alex-barstow) [reply src/js/player.js:3387]: Unfortunately I don't believe so. `retrying_` signals that at least one retry has been attempted, so we don't add the `removeRetryHandler` below more than once during any subsequent retries. At the point when we need to check whether to add that handler or not, `changingSrc_` is always `true` so it cannot serve the same purpose.

File: test/unit/player.test.js [hunk 3]
```diff
@@ -349,6 +349,89 @@ QUnit.test('should asynchronously fire error events during source selection', fu
...
+  // No more sources to try so the previous source should remain
+  player.trigger('error');
+
+  assert.deepEqual(
+    player.currentSource(),
+    { src: 'http://vjs.zencdn.net/v/oceans3.mp4', type: 'video/mp4' },
+    'last source remains'
+  );
+
+  player.dispose();
```
[the remark this answers is not in the document]
Comment (alex-barstow) [reply test/unit/player.test.js:394]: Good thinking

**Review by misteroneill — COMMENTED** [review]
There was an offline discussion about the issue of differentiating between a `src()` call that comes from a user vs. one that comes from the retry process.

One potential avenue for that is to add another argument to `src()` but I think a preferable approach would be to use a pseudo-private method to hide the second argument. Something like:
```js
src(source) {
  return this.src_(source);
}
src_(source, duringReset) {
  // Everything from current src() method, using duringReset as a 
  // differentiator and calling src_() internally.
}
```

File: test/unit/player.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (alex-barstow) [reply test/unit/player.test.js:404]: Done

File: src/js/player.js [hunk 4]
```diff
@@ -3377,6 +3392,48 @@ class Player extends Component {
...
+   *        algorithms can take the `type` into account.
+   *
+   *        If not provided, this method acts as a getter.
+   *
+   * @return {string|undefined}
+   *         If the `source` argument is missing, returns the current source
+   *         URL. Otherwise, returns nothing/undefined.
+   */
+  src(source) {
+    return this.handleSrc_(source);
```
Comment (gkatsev) [inline src/js/player.js:3436]: better to be explicit.
```suggestion
    return this.handleSrc_(source, false);
```

**Review by gkatsev — COMMENTED** [review]
Looks good. I did find an infinite error loop that's not great. Let me investigate it some more first.

File: src/js/player.js [hunk 5]
```diff
@@ -3377,6 +3392,48 @@ class Player extends Component {
...
+        this.off('error', retry);
+      };
+
+      this.one('error', retry);
+      // If we've already initiated a retry, we don't need to add this again
+      if (!isRetry) {
+        this.one('playing', stopListeningForErrors);
+      }
+
+      this.resetRetryOnError = () => {
```
Comment (gkatsev) [inline src/js/player.js:3413]:
```suggestion
      this.resetRetryOnError_ = () => {
```

File: src/js/player.js [hunk 6]
```diff
@@ -3342,7 +3344,17 @@ class Player extends Component {
...
 
-    this.cache_.sources = sources;
+    if (!isRetry) {
+      // Only update the cached source list if we are not retrying a new source after error,
+      // since in that case we want to include the failed source(s) in the cache
+      this.cache_.sources = sources;
+
+      // If a retry was previously started, reset it
+      if (this.resetRetryOnError) {
+        this.resetRetryOnError();
```
Comment (gkatsev) [inline src/js/player.js:3354]:
```suggestion
      if (this.resetRetryOnError_) {
        this.resetRetryOnError_();
```

**Comment by gkatsev** [comment]
So, the infinite loop happens in this type of cases:
```
player.src([
  {src: 'http://vjs.zencdn.net/v/oceans.foo', type: 'video/mp4'},
  {src: 'https://d2zihajmogu5jn.cloudfront.net/bipbop-advanced/bipbop_16x9_variant.m3u8', type:'application/x-mpegurl'},
  {src: 'http://vjs.zencdn.net/v/oceans.mp4', type: 'video/mp4'},
  ]);
player.on('error', () => {
  console.log('we errored');
  player.src({src:'http://vjs.zencdn.net/v/oceans.bar', type:'video/mp4'})
});
```

If I listen to the error with `one` instead of `on`, everything is fine. ~Interestingly, giving it a second source works fine.~ (seems to not be the case anymore?)
Looking at it some more, this would be an issue even without the retry on error. So, I don't think it's something we need to worry about here.

File: src/js/player.js (hunk 5 above)
Comment (alex-barstow) [reply src/js/player.js:3413]: Whoops good catch

**Comment by gkatsev** [comment]
So, with the following thing we end up with a player that works but I think it works because our error handlers aren't getting cleared up properly:
```js
player.src([
  {src: './oceans.foo', type: 'video/mp4'},
  {src: './oceans.mp4', type: 'video/mp4'},
  {src: 'https://d2zihajmogu5jn.cloudfront.net/bipbop-advanced/bipbop_16x9_variant.m3u8', type:'application/x-mpegurl'},
  ]);
player.on('error', () => {
  console.log('iteration %d', ++i, DomData.get(player.el()).handlers.error.length);
  player.src([
    {src:'./oceans.bar', type:'video/mp4'},
    {src:'./oceans.baz', type:'video/mp4'},
  {src: 'https://d2zihajmogu5jn.cloudfront.net/bipbop-advanced/bipbop_16x9_variant.m3u8', type:'application/x-mpegurl'},
  ]);
});
```
I then added this line below where we add the error handler
```js
      console.log('retry add', retry.guid)
```
and then the following before removal (in both playing handler and the reset method)
```js
      console.log('retry remove', retry.guid)
```
I then end up with the following output:
```
18:42:03.600 video.js:26905 retry add 551
18:42:03.891 video.js:26905 retry add 1019
18:42:03.892 video.js:26913 retry remove 1019
18:42:03.893 video.js:26905 retry add 1021
18:42:03.936 video.js:26913 retry remove 1021
18:42:03.937 video.js:26905 retry add 1034
18:42:03.939 video.js:26905 retry add 1037
18:42:03.954 video.js:26913 retry remove 1037
18:42:03.955 video.js:26905 retry add 1043
18:42:03.957 video.js:26905 retry add 1046
```
Notice that 551, 1034, and 1046 haven't been removed. I'd expect that only 1046 wouldn't have a matching remove yet.
When I hit play, it seems like all those pending ones get removed as expected:
```
18:42:19.803 video.js:26899 retry remove 551
18:42:19.804 video.js:26899 retry remove 1034
18:42:19.804 video.js:26899 retry remove 1043
```
I think these extra error handlers end up making the above snippet work.

**Comment by alex-barstow** [comment]
I updated the reset logic so we remove all retry-related event handlers on every `handleSrc_()` call, and this seems to fix the cleanup issues.

**Review by gkatsev — APPROVED** [review]
Verified locally that the handlers get removed properly now and double-checked how this works for autoplay.
LGTM!

**Comment by gkatsev** [comment]
This PR fixes #1805!

**Comment by John2397** [comment]
> This PR fixes #1805!

When can we expect the new release that will include this feature?

**Comment by gkatsev** [comment]
@John2397 this is available now in v7.12.

**Comment by gkatsev** [comment]
Looks like we forgot to document this option in https://github.com/videojs/video.js/blob/main/docs/guides/options.md

**Pull request merged**